airflow/www_rbac/filters.py

```diff
@@ -0,0 +1,95 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from flask import g
+from flask_appbuilder.models.sqla.filters import BaseFilter
+
+from airflow.www_rbac.app import appbuilder
+
+
+class AirflowFilter(BaseFilter):
+    """
+    Add utility function to make BaseFilter easy and fast
+
+    These utility function exist in the SecurityManager, but would do
+    a database round trip at every check. Here we cache the role objects
+    to be able to make multiple checks but query the db only once
+    """
+
+    def get_user_roles(self):
+        if g.user.is_anonymous():
+            public_role = appbuilder.config.get('AUTH_ROLE_PUBLIC')
+            return [appbuilder.sm.find_role(public_role)] if public_role else []
+        return g.user.roles
+
+    def get_all_permissions(self):
+        """
+        Returns a set of tuples with the perm name and view menu name
+        """
+        perms = set()
```

Comment:
My feeling around most of the methods in this new module is that they belong in the `User` class if anywhere outside of the `SupersetSecurityManager`.  This `set` is created for every assertion and then thrown away. I think in some cases the set will be create 3 times for a single page view.

You could change this to be a `get_and_cache_perms` method here that would build the `set` only once, but logically it fits better on the User model than on the BaseFilter class. Even though I it would work on this class also, because the Filter object's lifecycle in scope to the web request, it's unintuitive: it's confusing to think that something like `AirflowFilter._cache_perms` would be user-specific.